src/main/java/com/iota/iri/network/pipeline/TransactionProcessingPipeline.java

```diff
@@ -0,0 +1,356 @@
+        }, "pre-process-stage"));
+        stagesThreadPool.submit(new Thread(() -> {
+            try {
+                while (!shutdown.get()) {
+                    ProcessingContext ctx = validationStageQueue.take();
+                    ctx = validationStage.process(ctx);
+                    switch (ctx.getNextStage()) {
+                        case RECEIVED:
+                            receivedStageQueue.put(ctx);
+                            break;
+                        case MULTIPLE:
+                            ImmutablePair<ProcessingContext<ReplyPayload>,
+                                    ProcessingContext<
+                                            ReceivedPayload>> payload = (ImmutablePair<ProcessingContext<ReplyPayload>,
+                                                    ProcessingContext<ReceivedPayload>>) ctx.getPayload();
+                            replyStageQueue.put(payload.getLeft());
+                            receivedStageQueue.put(payload.getRight());
+                            break;
+                        case ABORT:
+                            break;
+                        default:
+                            // do nothing
+                    }
+                }
+            } catch (InterruptedException e) {
+            } finally {
+                log.info("validation stage shutdown");
+            }
+        }, "tx-validation-stage"));
+        stagesThreadPool.submit(new Thread(() -> {
+            try {
+                while (!shutdown.get()) {
+                    ProcessingContext ctx = replyStageQueue.take();
+                    replyStage.process(ctx);
+                }
+            } catch (InterruptedException e) {
+            } finally {
+                log.info("reply stage shutdown");
+            }
+        }, "reply-stage"));
+        stagesThreadPool.submit(new Thread(() -> {
+            try {
+                while (!shutdown.get()) {
+                    ProcessingContext ctx = receivedStageQueue.take();
+                    ctx = receivedStage.process(ctx);
+                    switch (ctx.getNextStage()) {
+                        case BROADCAST:
+                            broadcastStageQueue.put(ctx);
+                            break;
+                        default:
+                            // do nothing
+                    }
+                }
+            } catch (InterruptedException e) {
+            } finally {
+                log.info("received stage shutdown");
+            }
+        }, "received-stage"));
+        stagesThreadPool.submit(new Thread(() -> {
+            try {
+                while (!shutdown.get()) {
+                    ProcessingContext ctx = broadcastStageQueue.take();
+                    broadcastStage.process(ctx);
+                }
+            } catch (InterruptedException e) {
+            } finally {
+                log.info("broadcast stage shutdown");
+            }
+        }, "broadcast-stage"));
+    }
```

Comment:
Under this design each stage runs in a separate thread.
Maybe later some stages can be multithreaded?

Not sure if this is so important, but open an issue about it please

Reply:
This design doesn't hinder stages to be multi threaded later?